atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java

```diff
@@ -413,18 +412,15 @@ Set<CassandraServer> maybeFilterLocalHosts(Set<CassandraServer> hosts) {
         return hosts;
     }
 
-    private Optional<CassandraServer> getRandomHostByActiveConnections(Set<CassandraServer> desiredHosts) {
-
+    @VisibleForTesting
+    Optional<CassandraServer> getRandomHostByActiveConnections(Set<CassandraServer> desiredHosts) {
         Set<CassandraServer> localFilteredHosts = maybeFilterLocalHosts(desiredHosts);
-
-        Map<CassandraServer, CassandraClientPoolingContainer> matchingPools = KeyedStream.stream(
-                        ImmutableMap.copyOf(currentPools))
-                .filterKeys(localFilteredHosts::contains)
-                .collectToMap();
+        ImmutableMap<CassandraServer, CassandraClientPoolingContainer> matchingPools = currentPools.entrySet().stream()
```

Comment:
this will still allocate a `java.util.concurrent.ConcurrentHashMap.EntrySpliterator`, but we'll only iterate once instead of twice

Reply:
are we concerned about concurrent modification of the `currentPools` though? My understanding is that the previous implementation mitigated this by making a copy immediately, but now that we don't do that, the entrySet can change midway through?

Reply:
Although I suppose the concern about this `matchingPools` map changing _after_ construction (which was the original motivator behind the copyOf) is still safe as we collect to an immutable map.

Reply:
we have to iterate the currentPools once to either map the immediate snapshot copy or make the filtered copy, so I opted to to the one shot iterate & filter which would be more up-to-date in the sense of reflecting any concurrent additions/removals and avoids some intermediate overhead, and once we've constructed that snapshot of local cluster state, that's what we use to select the desired Cassandra target node.

Reply:
A stream or iterator on `ConcurrentHashMap.entrySet` will provide a weakly consistent view of the underlying concurrent map, so we will see all entries as they exist when stream is created and may see updates, but will not throw on concurrent modifications.

Per https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html#entrySet-- & https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/package-summary.html#Weakly
> Most concurrent Collection implementations (including most Queues) also differ from the usual java.util conventions in that their Iterators and Spliterators provide weakly consistent rather than fast-fail traversal:
> * they may proceed concurrently with other operations
> * they will never throw ConcurrentModificationException
> * they are guaranteed to traverse elements as they existed upon construction exactly once, and may (but are not guaranteed to) reflect any modifications subsequent to construction.

Reply:
fabulous - thanks for clearing that up